### PR TITLE
WIP Neuraxle refactor

### PR DESCRIPTION
**Pull request by alexbrillant**
- Use Neuraxle TensorflowV1ModelStep
- Use Dynamic RNN to predict sequence of dynamic lengths
- Clean tensorflow  code & placeholders

TODO: comments & docstrings...
Note : it wouldn't take long to migrate this version to tensorflow 2 ! 

How to use TrainingHelper and InferenceHelper together : 

https://stackoverflow.com/questions/49134432/how-to-use-tensorflow-seq2seq-without-embeddings

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    output = create_decoder_outputs(step, helper, encoder_state, decoder_cell)
+
+    return output
+
+
+def create_inference_decoder(step: TensorflowV1ModelStep, encoder_state, decoder_cell):
+    start_inputs = tf.constant(GO_TOKEN, shape=[step.hyperparams['batch_size'], step.hyperparams['output_dim']])
+
+    helper = tf.contrib.seq2seq.InferenceHelper(
```
Comment (alexbrillant) [inline model.py]: I am stuck here with the InferenceHelper.... At least, the TrainingHelper is training well ! Any ideas why the InferenceHelper is not working ? @guillaume-chevalier

Comment (alexbrillant) [reply model.py]: it works now.

Comment (alexbrillant) [reply model.py]: fixed with maximum iterations parameter

**Comment by guillaume-chevalier** [comment]
Here is what I think we could do with TF2: 

```
last_encoder_state, encoder_outputs = apply_encoder(encoder, tf_data_inputs)
last_encoder_output = encoder_outputs[:, -1, :]

stacked_cells = [gru_cell(...) for _ in range(stacked)]

rnn = RNN(stacked_cells)

decoder_outputs = []
state = last_encoder_state
for time_step in range(self.output_seq_len): 
    state, out = rnn(state_cell=state, input=last_encoder_output)  # repeat last encoder output as an input (cheap but very okay to do this here).
    decoder_outputs.append(out)

return decoder_outputs
```

This is some dirty pseudocode I've just done. For instance, the `apply_encoder` method is extracted in my code above as it seems not to be the problem we focus on. I think it would be possible to extract method as such for the decoder.

File: model.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    OUTPUT_SIZE = 5
+    EPOCHS = 150
+
+    def __init__(self):
+        super().__init__([
+            TensorflowV1ModelStep(
+                create_graph=create_graph,
+                create_loss=create_loss,
+                create_optimizer=create_optimizer,
+                create_feed_dict=create_feed_dict
```
Comment (guillaume-chevalier) [inline model.py]: What does this do? I need to understand everything

File: model.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    exercise = 4
+    # We choose which data function to use below, in function of the exericse.
+    if exercise == 1:
+        generate_x_y_data = generate_x_y_data_v1
+    if exercise == 2:
+        generate_x_y_data = generate_x_y_data_v2
+    if exercise == 3:
+        generate_x_y_data = generate_x_y_data_v3
+    if exercise == 4:
+        generate_x_y_data = generate_x_y_data_v4
```
Comment (guillaume-chevalier) [inline model.py]: Let's keep the 4 examples. Good.

File: model.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    pipeline = DeepLearningPipeline(
+        SignalPredictionPipeline(),
+        validation_size=0.15,
```
Comment (guillaume-chevalier) [inline model.py]: There was no validation in the original example. This seems like it is a problem here (and in the LSTM example, too).

File: model.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        batch_size=SignalPredictionPipeline.BATCH_SIZE,
+        batch_metrics={'mse': to_numpy_metric_wrapper(mean_squared_error)},
+        shuffle_in_each_epoch_at_train=True,
+        n_epochs=SignalPredictionPipeline.EPOCHS,
+        epochs_metrics={'mse': to_numpy_metric_wrapper(mean_squared_error)},
+        scoring_function=to_numpy_metric_wrapper(mean_squared_error)
+    )
+
+    data_inputs, expected_outputs = generate_x_y_data(isTrain=True, batch_size=SignalPredictionPipeline.BATCH_SIZE)
```
Comment (guillaume-chevalier) [inline model.py]: I'm not sure what happens here so the following may be wrong but I think I spotted a problem: 

I think your problem is that I was calling the `generate_x_y_data` function at each epoch. You'd probably need here a train and a test data generator class that lazy loads each batches for the DL Minibatching Pipeline. 

You'd also need to split the data on the time axis like this here if using cross validation: 


So the users could optimize on the validation data with a manual search and manual parameter tuning (no random search), and then call a function to test on the test set or something. So we need to probably change the `generate_x_y_data` method.

File: model.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    pipeline, outputs = pipeline.fit_transform(data_inputs, expected_outputs)
+
+    mse_train = pipeline.get_epoch_metric_train('mse')
+    mse_validation = pipeline.get_epoch_metric_validation('mse')
+
+    plot_metric(mse_train, mse_validation, xlabel='epoch', ylabel='mse', title='Model Mean Squared Error')
```
Comment (guillaume-chevalier) [inline model.py]: Could you please upload a preview of that plot / chart in your PR for me to inspect this without running the code locally yet?

File: requirements.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+tensorflow==1.14
+-e git://github.com/alexbrillant/Neuraxle.git@a270fe2b2f73c9350d76fcf4b6f058b764a8c8f7#egg=neuraxle
+requests
```
Comment (guillaume-chevalier) [inline requirements.txt]: @alexbrillant Let's change requests to `urllib`. Let's not import requests anymore. See how I use urllib here: 
https://www.neuraxle.org/stable/rest_api_serving.html#API-Call-Example

File: model.py (hunk 5 above)
Comment (alexbrillant) [reply model.py]: the cross validation is done in the deep learning pipeline.

Comment (alexbrillant) [reply model.py]: oh wait you don't have all my changes I actually have a loop that calls generate_x_y_date a few times :/

Comment (alexbrillant) [reply model.py]: let's change generate_x_y_data_v4 it was confusing

Comment (alexbrillant) [reply model.py]: also the normalization, and lazy data loading should be done in neuraxle
